### ISSUE TITLE
[iOS] Global create is open on account creation/sign-in, but isn't visible

### ISSUE THREAD

**Issue by JmillsExpensify** [Expensify/App#5303]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Create a new account
2. When you sign in, Global create is open but invisible
3. If you close Global Create and re-open, it shows up

<img width="270" alt="Screen Shot 2021-09-16 at 11 38 51 AM" src="https://user-images.githubusercontent.com/1127863/133666964-79a89f22-cd68-419f-a75a-df638a8efdc8.png">

## Expected Result:
Global create should be open and visible on account creation

## Actual Result:
Global create doesn't show up

## Workaround:
Workaround is close and re-open, but this is a critical first-user experience, so this workaround isn't acceptable

## Platform:

Where is this issue occurring?
- iOS

**Version Number:** 1.0.99-1
**Reproducible in staging?:** Yes
**Reproducible in production?:** Yes
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation
**Expensify/Expensify Issue URL:** N/A
**Issue reported by:** Marc
**Slack conversation:** https://expensify.slack.com/archives/C020EPP9B9A/p1631816718390200

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @ctkochan22 (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by ctkochan22** [comment]
Reproduced while testing fix for https://github.com/Expensify/Expensify/issues/177787#issuecomment-921377044

![image](https://user-images.githubusercontent.com/12980144/133713232-33eeb224-3453-4288-97bf-171df306b196.png)

Reproduction step:
1. Do not do anything on old dot. I just created a user via the clitools
2. Log into new dot ios directly

When i selected "buisiness" then "expensify card" via the welcome inbox tasks, this bug did not exist. I need to finish the other PR, and I'll re-iterate on these steps

**Comment by ctkochan22** [comment]
I bet it has something to do with why we need a timeout on this line: https://github.com/Expensify/App/blob/09e2d941c17e6cea48e93210be4e27fff9ece502/src/pages/home/sidebar/SidebarScreen.js#L84-L86

**Comment by ctkochan22** [comment]
@Beamanator do you have an idea? I see you are familiar with this area

**Comment by Beamanator** [comment]
I'm the one who merged in this `setTimeout` since I couldn't find a better solution 😅 

I created another issue to get contributor ideas why `setTimeout` was needed and how to replace it: https://github.com/Expensify/App/issues/5296

I'll try Marc's idea here: https://github.com/Expensify/App/pull/5214#issuecomment-920172507

**Comment by Beamanator** [comment]
Here's what happened when I tried @marcaaron 's idea, and the changes I made:

Workflow:

1. Open app on iOS, iPhone 12, version 14.5
2. Sign up with brand new account
3. Validate account using CLI
4. Log in with new account
5. Global create didn't open (no X either)

Code changed & logs added:

1. All code changes recommended here: https://github.com/Expensify/App/pull/5214#issuecomment-920172507
2. `NameValuePair.get()`:

```js
function get(name, onyxKey, defaultValue) {
    API.Get({
        returnValueList: 'nameValuePairs',
        name,
    })
        .then((response) => {
            const value = lodashGet(response.nameValuePairs, [name], defaultValue || '');
            // Log the time this is called so we know when Onyx.set "should have been" called
            if (name === 'isFirstTimeNewExpensifyUser') console.log('[AuthScreens] setting nvp to ', value);
            Onyx.set(onyxKey, value);
        });
}
```

3. `SidebarScreen.js`:
    - In `componentDidMount`, add log: `console.log('[SidebarScreen] componentDidMount');`
    - In `componentDidUpdate`, add log: `console.log(`[SidebarScreen] componentDidUpdate (prev:${prevProps.isFirstTimeNewExpensifyUser})(now:${this.props.isFirstTimeNewExpensifyUser})`);`
4. In `react-native-onyx/lib/Onyx.js`:
    - In `function set`, add log at the beginning: `if (key === 'isFirstTimeNewExpensifyUser') console.log('[Onyx.set]', key, val);`

Order of logs when following the workflow:

```
 LOG  [AuthScreens] setting nvp to  true
 LOG  [Onyx.set] isFirstTimeNewExpensifyUser true
 LOG  [SidebarScreen] componentDidMount
 LOG  [SidebarScreen] componentDidUpdate (prev:false)(now:false)
 LOG  [SidebarScreen] componentDidUpdate (prev:false)(now:false)
 ... repeated 7 more times
```

Question: In `SidebarScreen`'s `componentDidUpdate`, why is `this.props.isFirstTimeNewExpensifyUser` never `true`?? The log in `Onyx.set` shows it does get called with `true` so it should be updating and re-triggering `componentDidMount with the value `true`!!

**Comment by kidroca** [comment]
@Beamanator 
You get `false` in componentDidUpdate because `componentDidMount` sets it to `false` 

```js
if (this.props.isFirstTimeNewExpensifyUser) {
  /* Bunch of code */

  // Set the NVP to false so we don't automatically open the menu again
  // Note: this may need to be moved if this NVP is used for anything else later
  NameValuePair.set(CONST.NVP.IS_FIRST_TIME_NEW_EXPENSIFY_USER, false, ONYXKEYS.NVP_IS_FIRST_TIME_NEW_EXPENSIFY_USER);
}
```

So it's initially `true` during `componentDidMount`  and entered the above flow

---

I've deliberately altered this to force a `true` condition 

#### src/libs/actions/NameValuePair.js
```js
function get(name, onyxKey, defaultValue) {
    API.Get({
        returnValueList: 'nameValuePairs',
        name,
    })
        .then((response) => {
            let value = lodashGet(response.nameValuePairs, [name], defaultValue || '');

            if (onyxKey === 'isFirstTimeNewExpensifyUser') {
                console.log('isFirstTimeNewExpensifyUser: ', value);
                value = true;
            }

            Onyx.set(onyxKey, value);
        });
}
```

---

> Global create didn't open (no X either)

It's seems it might have to do with the other expression `const hasFreePolicy = _.chain(this.props.policies)` or however the code was changed. Maybe you can share push your local changes?

**Comment by Beamanator** [comment]
@kidroca yeah sorry my notes may not have been super clear, I ended up stripping most of the stuff in `componentDidMount` and moving it to `componentDidMount`, as suggested as a possible solution [here](https://github.com/Expensify/App/pull/5214#issuecomment-920172507)

Here's a branch with what I was testing: https://github.com/Expensify/App/tree/beaman-investigatingGlobalCreateFishyness

Note: As mentioned above, I also added a log in In `react-native-onyx/lib/Onyx.js`'s `set` function:
- Add log at the beginning: `if (key === 'isFirstTimeNewExpensifyUser') console.log('[Onyx.set]', key, val);`

**Comment by kidroca** [comment]
What I've found out is that without the `setTimeout` the modal won't appear (ios), a similar issue we had with the attachment picker 
It seems to be a known issue with iOS and Modals, but when I tried to recreated it in a standalone expo snack it didn't happen and the modal always appeared (I'm talking about the issue with the attachment picker).
So this might be something that we're doing as we have multiple component layers and HOCs until we render the Modal

**Comment by kidroca** [comment]
@Beamanator 
The code changes work for me in that they raise the `true` flag in the `componentDidUpdate` method
But due to the lack of `setTimeout` I only see the (x) button while the modal is invisible

One of the reasons you don't get the value in `componentDidUpdate` is because that's not the only `Onyx` connection that this component is using - the component won't mount until the keys `betas`, `policies` aren't read from storage as they don't have the `initWithStoredValues: false` setting. 
And my theory is this happens:
1. Backend responds before this component have mounted and sets the value to `true`. Now that's the value in storage
2. This component hasn't mounted because not all Onyx keys are ready
3. When the component mounts it does not use the value stored in Onyx - due to `initWithStoredValues: false`
4. When `componentDidUpdate` runs you see the `defaultProps.isFirstTimeNewExpensifyUser` which is `false`

**Comment by kidroca** [comment]
Here's a non `setTimeout` based solution, though it's against the Expensify, "no Promise" policy, but I think it shows a use case where it's favorable to have the option to use Promises

Make `NameValuePair` methods async

```jsx
/* we want to open the menu for first time experience, and then we flag the experience as complete */
NameValuePair
    .get(CONST.NVP.IS_FIRST_TIME_NEW_EXPENSIFY_USER, ONYXKEYS.NVP_IS_FIRST_TIME_NEW_EXPENSIFY_USER, true)
    .then(value => this.setState({isCreateMenuActive: value}))
    .then(() => NameValuePair.set(CONST.NVP.IS_FIRST_TIME_NEW_EXPENSIFY_USER, false, ONYXKEYS.NVP_IS_FIRST_TIME_NEW_EXPENSIFY_USER));
```

Here's a full diff:
https://github.com/Expensify/App/compare/main...kidroca:kidroca/global-create-fishyness
This way handling is much simple and runs only once. You don't have to setup a check that runs for every `componentDidMount`

https://user-images.githubusercontent.com/12156624/133812154-718f4f9c-cf40-4466-9026-e89a4faa2622.mov

**Comment by kidroca** [comment]
To prove there's some weird bug with the Modal on iOS you can force the Modal to be open right from the start 

#### src/pages/home/sidebar/SidebarScreen.js
```jsx
        this.state = {
            isCreateMenuActive: true,
        };
```

Refresh and you'll end up with this:
![image](https://user-images.githubusercontent.com/12156624/133815643-29a0b3b9-77eb-45ad-9f62-b90f3a7781ef.png)

**Comment by ctkochan22** [comment]
Ooo interesting. Let me test this out locally (sorry just catching up)

**Comment by ctkochan22** [comment]
Still trying to grasp the issue. If i'm following your logic correctly, @kidroca , you're saying it doesn't run componentDidMount() until the ONYX keys are loaded. So it defaults to `defaultProps.isFirstTimeNewExpensifyUser` which is not set yet?

**Comment by marcaaron** [comment]
I think Peter is saying this isn't a React lifecycle problem, but maybe a bug with `react-native-modal` on iOS? i.e. it still won't open even with a hardcoded value of `isVisible={true}`.

**Comment by kidroca** [comment]
> If i'm following your logic correctly, @kidroca , you're saying it doesn't run componentDidMount() until the ONYX keys are loaded.

Correct.
This is only a problem with the code that sets `initWithStoredValues: false` and moves the logic to  `componentDidUpdate`

To make that work correctly we'll want to remove `initWithStoredValues: false` and handle the `isFirstTimeNewExpensifyUser` in both `didMount` and `didUpdate`, and even then we'll probably still have the problem

---

> 2. When you sign in, Global create **is open** but invisible

Judging by the problem description, the "existing" logic is working correctly, it's just that it "fires" too soon (even with a timeout)
as when you hardcode it the modal to open: https://github.com/Expensify/App/issues/5303#issuecomment-921893532

**Comment by parasharrajat** [comment]
> I think Peter is saying this isn't a React lifecycle problem, but maybe a bug with react-native-modal on iOS? i.e. it still won't open even with a hardcoded value of isVisible={true}.

I don't think it's true. https://snack.expo.dev/@parasaharrajat/react-native-modal-example

Check the Snack. It works great.

**Comment by kidroca** [comment]
> I think Peter is saying this isn't a React lifecycle problem, but maybe a bug with `react-native-modal` on iOS? i.e. it still won't open even with a hardcoded value of `isVisible={true}`.

Yes, I've seen others have the problem opening a modal as soon as another modal closes, which results in a similar experience - `isPopoupVisible` would be `true` while there would be no modal on screen
But I couldn't recreate it in a standalone environment to submit a bug for `react-native-modal` so it might be on us

**Comment by puneetlath** [comment]
I just tried this flow while testing and the global create modal was open for me by default on iOS v1.0.99.2 FYI. 

![IMG_2249](https://user-images.githubusercontent.com/1371996/133833992-598a14d7-b240-4fce-bf23-d90a0a550e77.PNG)

**Comment by kidroca** [comment]
Yes, the current logic works -  for some users the delay from `setTimout` is enough - no issue, for other it's not - bug, the modal does not appear

**Comment by JmillsExpensify** [comment]
Yeah in testing I've gotten global create to open flawlessly, but then others not. It's just still spotty.

**Comment by nhanth1012** [comment]
@JmillsExpensify 
This issue is still on working?
May I try to find the solution now?

**Comment by roryabraham** [comment]
No substantial update from me – I tried upgrading `@react-navigation/drawer` to `^6.1.4` and that did not fix the issue. I also confirmed that the `isVisible` prop of the `ReactNativeModal` component is successfully being changed from `false` to `true`, so I don't think that this is really a react lifecycle issue. Still don't know why the modal is not visible, nor do I understand what race condition is at play to make it sometimes-fixable via `setTimeout`.

**Comment by kidroca** [comment]
I've made a couple of more experiments that reduce the scope to using a `Modal` here: 
https://github.com/Expensify/App/blob/df229d1f7c6933ce5efcfde4b77ddf89c219acbb/src/libs/Navigation/AppNavigator/MainDrawerNavigator.js#L49

I've imported `react-native-modal` directly and replaced the `<PopoverMenu />` with `<Modal isModalVisible={true}>{/* some sample content */}</Modal>` and the modal would still fail to render.
A modal can be opened only a "some time" after the `drawerContent` (SidebarScreen) render
Adding `<Modal isModalVisible={true}>{/* some sample content */}</Modal>` anywhere else in the app pop ups a modal when you restart the app. adding it to `drawerContent` does not popup the modal when you restart the app

The `drawerContent` receives a `progress` prop: https://reactnavigation.org/docs/drawer-navigator#drawercontent
Maybe we have to make sure the progress has settled before opening the Modal

I can try to recreated this in a snack and submit an issue with RN navigation if it's recreatable

**Comment by parasharrajat** [comment]
@kidroca I spent some time yesterday figuring out this. It seems that isVisible prop is not being passed to the Modal for Popovermenu on load. It was strange experience. I debugged Modal component and props.visible was never true for this case. May be yes it has to do with Drawer but I haven't looked at that yet.

**Comment by kidroca** [comment]
@parasharrajat 
> @kidroca I spent some time yesterday figuring out this. It seems that isVisible prop is not being passed to the Modal for Popovermenu on load. It was strange experience. I debugged Modal component and props.visible was never true for this case. May be yes it has to do with Drawer but I haven't looked at that yet.

It is being passed, there are other modals that are rendered besides the one we're discussing, you'll be getting `isVisible: false` because of them. So if you're just logging something  (or breaking in the debugging) out from `src/components/Modal/BaseModal.js` it would of course log `false` for the closed modals

I've just double checked and `isVisible` is passed correctly to the `BaseModal`

---

The issue is not related to what we pass, don't pass or how we setup the Modal, because even when you replace the `<PopoverMenu />` with a `react-native-modal` hardcoded to `isVisible` the modal would not show at all

**Comment by parasharrajat** [comment]
I made sure to test this scene, but `react-native-modal` instance is not receiving `isVisible:true` for the FAbMenu modal on load. Yeah, `baseModal` is correctly seeing true for isVisible prop. 
Then I directly passed true to `react-native-modal` instance in BaseModal but isVisible was not passed `true` for a single render.

I tried reducing the render to minimal but I found that it's not affected by rerendering.

**Comment by roryabraham** [comment]
I saw the `isVisible` prop being passed correctly to the `react-native-modal` from the `SidebarScreen` -> `PopoverMenu` -> `BasePopoverMenu` -> `ReactNativeModal`. So I think @kidroca is correct that `isVisible` is being passed correctly and that there's some race condition at play with the `drawerContent`.

**Comment by roryabraham** [comment]
Tried simply wrapping the `SidebarScreen` with `DrawerContentScrollView` as suggested by the docs [here](https://reactnavigation.org/docs/drawer-navigator/#providing-a-custom-drawercontent) but that didn't solve the problem.

**Comment by parasharrajat** [comment]
I ran a debugger over the `react-native-modal` and the logic inside never ran where we are checking `props.isvisible` is true in the onLoad scene.

I removed the nested props and directly set the isVisible true in BaseModal and the result was the same.

I am not what could be the reason that reacts is not passing that prop correctly.

**Comment by roryabraham** [comment]
> The drawerContent receives a progress prop: reactnavigation.org/docs/drawer-navigator#drawercontent
Maybe we have to make sure the progress has settled before opening the Modal

Was trying this, but ran into an issue where this prop just isn't present: https://github.com/react-navigation/react-navigation/issues/9878

**Comment by roryabraham** [comment]
In the meantime, can we maybe fix this more consistently by just increasing the timeout to 1500 ms or something? AFAICT that wouldn't result in a very adverse effect in the UX – when you open the app for the first time, 1.5 seconds later the global create menu opens. This is prettymuch the effect we want, and the minor delay likely isn't going to have an effect on usability. This one-time action is not a performance-critical flow, so the hack might be acceptable. Thoughts?

**Comment by kidroca** [comment]
> In the meantime, can we maybe fix this more consistently by just increasing the timeout to 1500 ms or something?

I think if we're in a hurry and it wont hurt UX, it'll be ok for the moment
I've tried different things, but nothing helped, like:
- removing `mode="modal"` from `AuthScreens` 
- adding `detachInactiveScreens={false}` to `BaseDrawerNavigator` 

We should at least change the logic for the check to be **inside** the `setInterval` e.g.

```jsx
    componentDidMount() {
        Performance.markStart(CONST.TIMING.SIDEBAR_LOADED);
        Timing.start(CONST.TIMING.SIDEBAR_LOADED, true);

        setTimeout(() => {
            if (this.props.isFirstTimeNewExpensifyUser) {
                 /* rest of the code */
            }
        });
    }
```

This way we're using the latest `isFirstTimeNewExpensifyUser` value, and not the one from "mount time" which might not have been set by the response

**Comment by parasharrajat** [comment]
Yeah, an average delay should be added as it does not feel good when using the first time, active modal was shown instead it should animate before the user.

**Comment by Jag96** [comment]
> I think if we're in a hurry and it wont hurt UX, it'll be ok for the moment

Agreed, this seems like the easiest fix right now

> We should at least change the logic for the check to be inside the setInterval

+1 as well

> Yeah, an average delay should be added as it does not feel good when using the first time, active modal was shown instead it should animate before the user with an instruction that what it does.

I agree adding some onboarding indications could make the UX better here, but I think the 1500ms will work for now.

**Comment by roryabraham** [comment]
Noticing a lot of people are assigned to this issue, but I'm going to submit a pull request to increase the timeout. Please DM me if you're already working on this.

**Comment by MitchExpensify** [comment]
No closed Botify

**Comment by roryabraham** [comment]
Why no closed?

**Comment by MitchExpensify** [comment]
I thought we closed when the PR hit production

**Comment by roryabraham** [comment]
For external contributors we close the issue 7 days after the fix has been deployed to production. For internal contributors we generally close it right away (and reopen if the PR fails QA)

**Comment by marcaaron** [comment]
I'm trying to understand this issue a bit to clear up the `setTimeout()` but not able to reproduce the bug anymore.

I can see it happen if we try to initialize the global create menu with `isActive: true`, but if we set to state after the component mounts then everything works fine.

Maybe it only affects certain versions of iOS...

**Comment by marcaaron** [comment]
Tested on iOS 13-15 and can't get it to happen. I think the moving the logic in [this PR](https://github.com/Expensify/App/pull/7865/files) fixed the problem, but unsure why.

**Comment by luacmartins** [comment]
It was definitely still happening then, because I tried to remove it and failed. Although I thought I left a comment somewhere about it fixing a race condition and not a bug with react-navigation.

**Comment by marcaaron** [comment]
>Initializing with isActive: false and then changing it - sometimes work

Interesting, I'm trying to find out what the "sometimes" factor is here. So far, it has been working fine for me and anyone else who has tested on iOS so I'm inclined to see if it will pop up again.

**Comment by marcaaron** [comment]
Changes here are what I have been testing -> https://github.com/Expensify/App/pull/8827
